python/ray/data/dataset.py

```diff
@@ -993,29 +996,37 @@ def union(self, *other: List["Dataset[T]"]) -> "Dataset[T]":
         """
 
         start_time = time.perf_counter()
-        context = DatasetContext.get_current()
-        tasks: List[ReadTask] = []
-        block_partition_refs: List[ObjectRef[BlockPartition]] = []
-        block_partition_meta_refs: List[ObjectRef[BlockPartitionMetadata]] = []
 
         datasets = [self] + list(other)
+        bls = []
+        has_nonlazy = False
         for ds in datasets:
             bl = ds._plan.execute()
-            if isinstance(bl, LazyBlockList):
+            if not isinstance(bl, LazyBlockList):
+                has_nonlazy = True
+            bls.append(bl)
+        if has_nonlazy:
+            blocks = []
+            metadata = []
+            for bl in bls:
+                if isinstance(bl, LazyBlockList):
+                    bs, ms = bl._get_blocks_with_metadata()
+                else:
+                    bs, ms = bl._blocks, bl._metadata
+                blocks.extend(bs)
+                metadata.extend(ms)
+            blocklist = BlockList(blocks, metadata)
```

Comment:
This block removes the previous shoe-horning of non-lazy block lists into lazy block lists with dummy read tasks. This now means that if a single dataset in a union has a non-lazy data source, then all remaining read tasks for all datasets with lazy data sources will be triggered.